sources/api/bootstrap-containers/src/main.rs

```diff
@@ -0,0 +1,567 @@
+            Ok(()) => Ok(true),
+            Err(e) => {
+                // If the systemd unit is not enabled, then `systemctl is-enabled` will return a
+                // non-zero exit code.
+                match e {
+                    error::Error::CommandFailure { .. } => Ok(false),
+                    _ => {
+                        // Otherwise, we return the error
+                        Err(e)
+                    }
+                }
+            }
+        }
+    }
+
+    fn is_active(&self) -> Result<bool> {
+        match command(SYSTEMCTL_BIN, &["is-active", &self.unit]) {
+            Ok(()) => Ok(true),
+            Err(e) => {
+                // If the systemd unit is not active(running), then `systemctl is-active` will
+                // return a non-zero exit code.
+                match e {
+                    error::Error::CommandFailure { .. } => Ok(false),
+                    _ => {
+                        // Otherwise, we return the error
+                        Err(e)
+                    }
+                }
+            }
+        }
+    }
+
+    fn enable(&self) -> Result<()> {
+        // Only enable the unit, since it will be started once systemd reaches the `configured`
+        // target
+        command(SYSTEMCTL_BIN, &["enable", &self.unit])
+    }
+
+    fn disable_and_stop(&self) -> Result<()> {
+        command(SYSTEMCTL_BIN, &["disable", &self.unit, "--now"])
+    }
+}
+
+/// Wrapper around process::Command that adds error checking.
+fn command<I, S>(bin_path: &str, args: I) -> Result<()>
+where
+    I: IntoIterator<Item = S>,
+    S: AsRef<OsStr>,
+{
+    let mut command = Command::new(bin_path);
+    command.args(args);
+    let output = command
+        .output()
+        .context(error::ExecutionFailure { command })?;
+
+    trace!("stdout: {}", String::from_utf8_lossy(&output.stdout));
+    trace!("stderr: {}", String::from_utf8_lossy(&output.stderr));
+
+    ensure!(
+        output.status.success(),
+        error::CommandFailure { bin_path, output }
+    );
+    Ok(())
+}
+
+/// Handles the `create-containers` subcommand
+async fn create_containers<P>(socket_path: P) -> Result<()>
+where
+    P: AsRef<Path>,
+{
+    let mut failed = 0usize;
+    let bootstrap_containers = get_bootstrap_containers(socket_path).await?;
+    for (name, container_details) in bootstrap_containers.iter() {
+        // Continue to handle other bootstrap containers if we fail one
+        if let Err(e) = handle_bootstrap_container(name, container_details) {
+            failed += 1;
+            error!("Failed to handle bootstrap container '{}': {}", &name, e);
+        }
+    }
+
+    ensure!(
+        failed == 0,
+        error::ManageContainersFailed {
+            failed,
+            tried: bootstrap_containers.len()
+        }
+    );
+
+    Ok(())
+}
+
+/// Handles the `mark-bootstrap` subcommand
+async fn mark_bootstrap(args: Args) -> Result<()> {
```

Comment:
```suggestion
async fn mark_bootstrap(args: MarkBootstrapArgs) -> Result<()> {
```

I think it would be better if all of the argument parsing were done in dedicated argument parsing functions.

The pattern to look at would be `apiclient` where we have

```rust
enum Subcommand {
    MarkBootstrap(MarkBootstrapArgs),
    CreateContainers(CreateContainersArgs),
}
```

Reply:
I updated the mechanism that I used to get command line arguments, to be similar to your proposal.